plenum/test/node_catchup/test_process_catchup_replies.py

```diff
@@ -0,0 +1,156 @@
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply2)
+
+    reply3 = catchup_reps[2]
+    ledger_manager.processCatchupRep(reply3, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply2)
+
+    reply1 = catchup_reps[0]
+    ledger_manager.processCatchupRep(reply1, sdk_wallet_client[1])
+    check_replies_applied(old_ledger_size,
+                          ledger,
+                          ledger_info,
+                          [reply1, reply2, reply3])
+    old_ledger_size = ledger.size
+
+    # send replies in next order: 6, 4, 5
+    # and check that after sending reply 4, it will be applied.
+    # Check that after sending reply 5, replies 5 and 6 will be applied.
+    reply6 = catchup_reps[5]
+    ledger_manager.processCatchupRep(reply6, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply6)
+
+    reply4 = catchup_reps[3]
+    ledger_manager.processCatchupRep(reply4, sdk_wallet_client[1])
+    check_replies_applied(old_ledger_size, ledger, ledger_info, [reply4])
+    old_ledger_size = ledger.size
+
+    reply5 = catchup_reps[4]
+    ledger_manager.processCatchupRep(reply5, sdk_wallet_client[1])
+    check_replies_applied(old_ledger_size, ledger, ledger_info, [reply5, reply6])
+
+    assert not ledger_info.receivedCatchUpReplies
+
+
+def test_process_invalid_catchup_reply(txnPoolNodeSet, looper, sdk_wallet_client):
+    '''
+    Test correct work of method processCatchupRep and that sending replies
+    in reverse order will call a few iterations of cycle in _processCatchupRep
+    '''
+    # count of transactions and replies (one txn in one reply)
+    # and iterations in _processCatchupRep
+    ledger_manager = txnPoolNodeSet[0].ledgerManager
+    ledger_info = ledger_manager.getLedgerInfoByType(ledger_id)
+    ledger = ledger_manager.ledgerRegistry[ledger_id].ledger
+    old_ledger_size = ledger.size
+    num_txns_in_reply = 3
+    reply_count = 1
+
+    cons_proof, catchup_reps = _add_txns_to_ledger(txnPoolNodeSet[1],
+                                                   looper,
+                                                   sdk_wallet_client,
+                                                   num_txns_in_reply,
+                                                   reply_count)
+    ledger_info.catchUpTill = cons_proof
+    ledger_info.state = LedgerState.syncing
+
+    # make invalid catchup reply by dint of adding new transaction in it
+    reply = catchup_reps[0]
+    txns = getattr(reply, f.TXNS.nm)
+    req = sdk_signed_random_requests(looper, sdk_wallet_client, 1)[0]
+    txns[str(old_ledger_size + 4)] = append_txn_metadata(reqToTxn(req), txn_time=12345678)
+    ledger_manager.processCatchupRep(CatchupRep(ledger_id,
+                                                txns,
+                                                getattr(reply, f.CONS_PROOF.nm)),
+                                     sdk_wallet_client[1])
```

Comment:
It makes sense to verify the behavior when an invalid reply is processed after some valid ones have been processed. E.g. `catchup_reps[1]` is invalid and is received first and `catchup_reps[0]` is valid and is received second.

Reply:
Done